packages/api/src/controllers/stream.test.ts

```diff
@@ -524,6 +524,76 @@ describe("controllers/stream", () => {
         expect(server.db.stream.addDefaultFields(document)).toEqual(stream);
       });
 
+      it("should lock pull for a new stream", async () => {
+        // Create stream pull
+        const now = Date.now();
+        const res = await client.put("/stream/pull", postMockPullStream);
+        expect(res.status).toBe(201);
+        const stream = await res.json();
+
+        // Request pull lock
+        const resLockPull = await client.post(`/stream/${stream.id}/lockPull`);
+        expect(resLockPull.status).toBe(204);
+
+        // Check that the lockedAt is marked with the correct date
+        const res2 = await client.get(`/stream/${stream.id}`);
+        expect(res2.status).toBe(200);
+        const stream2 = await res2.json();
+        expect(stream2.pull.lockedAt).toBeGreaterThan(now);
+      });
+
+      it("should not lock pull for an active stream", async () => {
+        // Create stream pull
+        const res = await client.put("/stream/pull", postMockPullStream);
+        expect(res.status).toBe(201);
+        const stream = await res.json();
+
+        // Mark stream as active
+        await db.stream.update(stream.id, { isActive: true });
+
+        // Requesting pull lock should fail, because the stream is active (so it should be replicated instead of being pulled)
+        const reslockPull = await client.post(`/stream/${stream.id}/lockPull`);
+        expect(reslockPull.status).toBe(423);
+      });
+
+      it("should not lock pull for already locked pull", async () => {
+        // Create stream pull
+        const res = await client.put("/stream/pull", postMockPullStream);
+        expect(res.status).toBe(201);
+        const stream = await res.json();
+
+        // Request pull lock
+        const resLockPull = await client.post(`/stream/${stream.id}/lockPull`);
+        expect(resLockPull.status).toBe(204);
+
+        // Requesting pull lock should fail, because pull has already locked b y another node (so it should be replicated instead of being pulled)
+        const resLockPull2 = await client.post(`/stream/${stream.id}/lockPull`);
+        expect(resLockPull2.status).toBe(423);
```

Comment:
Could be fun to make these in parallel with a `Promise.all` to actually make sure there are no races.
e.g.
```suggestion
        const promises = [];
        for (int i = 0; i < 10; i++) {
          promises.push(client.post(`/stream/${stream.id}/lockPull`));
        }
        const resPulls = await Promise.all(promises);
        expect(resPulls.filter(r => r.status === 204).length).toBe(1);
        expect(resPulls.filter(r => r.status === 423).length).toBe(9);
```
(sounded too fun I had to try. not sure it even "compiles" tho lol)

Reply:
Good idea, updated.